examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
+                    // If the migrated model already supports the data format, go ahead with the migration.
+                    transformedData = exportedData;
+                } else if (this.dataTransformationCallback !== undefined) {
+                    // Otherwise, try using the dataTransformationCallback if provided to get the exported data into
+                    // a format that we can import.
+                    try {
+                        transformedData = await this.dataTransformationCallback(exportedData, migratedModel.version);
+                    } catch {
+                        // TODO: This implies that the contract is to throw if the data can't be transformed, which
+                        // isn't great.  How should the dataTransformationCallback indicate failure?
+                        this.emit("migrationNotSupported", acceptedVersion);
+                        this._migrationP = undefined;
+                        return;
+                    }
+                } else {
+                    // We can't get the data into a format that we can import, give up.
+                    this.emit("migrationNotSupported", acceptedVersion);
+                    this._migrationP = undefined;
+                    return;
+                }
+                await migratedModel.importData(transformedData);
 
-            // TODO: Support retry
-            // TODO: Use TaskManager here to reduce container noise.  Specifically -- all clients should race up to
-            // this point (so they're all prepared to do the migration) but should wait for the task lock before
-            // attempting the attach() to minimize the chance that multiple containers are created on the service.
-            const containerId = await createResponse.attach();
+                // Store the detached model for later use and retry scenarios
+                this._preparedDetachedModel = detachedModel;
+            };
 
-            // Again, it could be the case that someone else finished the migration during our attach.
-            // Casting to MigrationState because TS doesn't understand that the state may have changed during the
-            // above await.
-            if (migratable.migrationTool.migrationState as MigrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
+            const completeTheMigration = async () => {
+                assert(this._preparedDetachedModel !== undefined, "this._preparedDetachedModel should be defined");
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                // Volunteer to complete the migration.
+                let isAssigned: boolean;
+                try {
+                    isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
```

Comment:
Does volunteerForMigration acquires lock?

Reply:
It will try to acquire the lock yes. If it does then `isAssigned` will be true, and if not then it will return false (another client completed it), or we will throw an error due to a disconnection.